File: Data Structures & Algorithms/majority-element/submission-1.py
class Solution:

    def majorityElement(self, nums):

        fq = 0
        ans = 0

        for i in range(len(nums)):

            if fq == 0:
                ans = nums[i]

            if ans == nums[i]:
                fq += 1
            else:
                fq -= 1

        count = 0

        for val in nums:
            if val == ans:
                count += 1

        return ans if count > len(nums) // 2 else -1

File: Data Structures & Algorithms/find-lucky-integer-in-an-array/submission-1.cpp
// class Solution {
// public:
//     int findLucky(vector<int>& arr) {
//         unordered_map<int, int> freq;

//         for (int num : arr) {
//             freq[num]++;
//         }

//         int ans = -1;

//         for (auto &p : freq) {
//             if (p.first == p.second) {
//                 ans = max(ans, p.first);
//             }
//         }

//         return ans;
//     }
// };



// //// 2nd

// class Solution {
// public:
//     int findLucky(vector<int>& arr) {
//         unordered_map<int, int> freq;

//         for (int num : arr)
//             ++freq[num];

//         int ans = -1;

//         for (const auto& [num, count] : freq)
//             if (num == count)
//                 ans = max(ans, num);

//         return ans;
//     }
// };


///// 3rd

class Solution {
public:
    int findLucky(vector<int>& arr) {
        int freq[501] = {0};

        for (int num : arr)
            ++freq[num];

        for (int i = 500; i >= 1; --i) {
            if (freq[i] == i)
                return i;
        }

        return -1;
    }
};